src/server/worker/datum_iterator_test.go

```diff
@@ -102,18 +103,135 @@ func TestDatumIterators(t *testing.T) {
+	c := getPachClient(b)
+	defer require.NoError(b, c.DeleteAll())
+	require.NoError(b, activateEnterprise(c))
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		dataRepo := tu.UniqueString("TestDatumIteratorPFS_data")
+		require.NoError(b, c.CreateRepo(dataRepo))
+
+		// put files in structured in a way so that there are many ways to glob it
+		commit, err := c.StartCommit(dataRepo, "master")
+		require.NoError(b, err)
+		for i := 0; i < 100*j; i++ {
+			_, err = c.PutFile(dataRepo, commit.ID, fmt.Sprintf("foo%v", i), strings.NewReader("bar"))
+			require.NoError(b, err)
+		}
+
+		require.NoError(b, err)
+		require.NoError(b, c.FinishCommit(dataRepo, commit.ID))
+
+		// make one with zero datums for testing edge cases
+		in0 := client.NewPFSInput(dataRepo, "!(**)")
+		in0.Pfs.Commit = commit.ID
+		pfs0, err := NewDatumIterator(c, in0)
+		require.NoError(b, err)
+
+		in1 := client.NewPFSInput(dataRepo, "/foo?1*")
+		in1.Pfs.Commit = commit.ID
+		pfs1, err := NewDatumIterator(c, in1)
+		require.NoError(b, err)
+
+		in2 := client.NewPFSInput(dataRepo, "/foo*2")
+		in2.Pfs.Commit = commit.ID
+		pfs2, err := NewDatumIterator(c, in2)
+		require.NoError(b, err)
+
+		in3 := client.NewUnionInput(in1, in2)
+		union1, err := NewDatumIterator(c, in3)
+		require.NoError(b, err)
+
+		in4 := client.NewCrossInput(in1, in2)
+		cross1, err := NewDatumIterator(c, in4)
+		require.NoError(b, err)
+
+		in5 := client.NewCrossInput(in3, in4)
+		cross2, err := NewDatumIterator(c, in5)
+		require.NoError(b, err)
+
+		// cross with a zero datum input should also be zero
+		in6 := client.NewCrossInput(in3, in0, in2, in4)
+		cross3, err := NewDatumIterator(c, in6)
+		require.NoError(b, err)
+
+		// zero cross inside a cross should also be zero
+		in7 := client.NewCrossInput(in6, in1)
+		cross4, err := NewDatumIterator(c, in7)
+		require.NoError(b, err)
+
+		in8 := client.NewPFSInputOpts("", dataRepo, "", "/foo(?)(?)*", "$1$2", false)
+		in8.Pfs.Commit = commit.ID
+		in9 := client.NewPFSInputOpts("", dataRepo, "", "/foo(?)(?)*", "$2$1", false)
+		in9.Pfs.Commit = commit.ID
+
+		join1, err := newJoinDatumIterator(c, []*pps.Input{in8, in9})
+
+		validateDI(b, pfs0)
+		validateDI(b, pfs1)
+		validateDI(b, pfs2)
+		validateDI(b, cross1)
+		validateDI(b, cross2)
+		validateDI(b, cross3)
+		validateDI(b, cross4)
+		validateDI(b, union1)
+		validateDI(b, join1)
+	}
 }
 
-func validateDI(t *testing.T, di DatumIterator, datums ...string) {
+func BenchmarkDI1(b *testing.B)  { benchmarkDatumIterators(1, b) }
+func BenchmarkDI5(b *testing.B)  { benchmarkDatumIterators(5, b) }
+func BenchmarkDI10(b *testing.B) { benchmarkDatumIterators(10, b) }
+func BenchmarkDI15(b *testing.B) { benchmarkDatumIterators(15, b) }
+func BenchmarkDI20(b *testing.B) { benchmarkDatumIterators(20, b) }
+func BenchmarkD25(b *testing.B)  { benchmarkDatumIterators(25, b) }
+
+func validateDI(t testing.TB, di DatumIterator, datums ...string) {
 	i := 0
+	clone := di
 	for di.Next() {
```

Comment:
I think this benchmark should also try calling `DatumN` with an n that isn't what the iterator is currently pointing to. I think that's the code that has the most low hanging fruit in terms of optimization. It's also going to pretty directly correspond to job performance, because workers will start by calling `DatumN` to get to the datums they need to process.